File: kotlin/src/main/kotlin/SnapkitImageURL.kt
package dev.snapkit.imageurl

import android.net.Uri

/**
 * Image transformation parameters
 */
data class TransformOptions(
    /** Image width (pixels) */
    val w: Int? = null,
    /** Image height (pixels) */
    val h: Int? = null,
    /** Resize method */
    val fit: Fit? = null,
    /** Output format */
    val format: Format? = null,
    /** Rotation angle (degrees) */
    val rotation: Int? = null,
    /** Blur strength (0.3-1000) */
    val blur: Int? = null,
    /** Whether to convert to grayscale */
    val grayscale: Boolean? = null,
    /** Whether to flip vertically */
    val flip: Boolean? = null,
    /** Whether to flip horizontally */
    val flop: Boolean? = null,
    /** Region extraction */
    val extract: Extract? = null,
    /** Device Pixel Ratio (1.0-4.0) */
    val dpr: Double? = null,
    /** Image quality (1-100) */
    val quality: Int? = null
) {
    /** Resize method */
    enum class Fit {
        CONTAIN, COVER, FILL, INSIDE, OUTSIDE;

        override fun toString() = name.lowercase()
    }

    /** Output format */
    enum class Format {
        JPEG, PNG, WEBP, AVIF;

        override fun toString() = name.lowercase()
    }

    /** Region extraction */
    data class Extract(
        val x: Int,
        val y: Int,
        val width: Int,
        val height: Int
    )
}

/**
 * Snapkit image URL builder
 *
 * @property organizationName Organization name (used as Snapkit subdomain)
 *
 * @example
 * ```kotlin
 * // S3 Direct Access (Recommended)
 * val builder = SnapkitImageURL("my-org")
 * val imageUrl = builder.build(
 *     path = "project/images/hero.jpg",
 *     transform = TransformOptions(
 *         w = 300,
 *         h = 200,
 *         fit = TransformOptions.Fit.COVER,
 *         format = TransformOptions.Format.WEBP
 *     )
 * )
 * // → "https://my-org-cdn.snapkit.studio/project/images/hero.jpg?transform=w:300,h:200,fit:cover,format:webp"
 * ```
 *
 * @example
 * ```kotlin
 * // External CDN Proxy (Optional)
 * val imageUrl = builder.build(
 *     url = "https://cdn.cloudfront.net/image.jpg",
 *     transform = TransformOptions(
 *         w = 300,
 *         h = 200,
 *         fit = TransformOptions.Fit.COVER,
 *         format = TransformOptions.Format.WEBP
 *     )
 * )
 * // → "https://my-org.snapkit.dev/image?url=https%3A%2F%2F...&transform=w:300,h:200,fit:cover,format:webp"
 * ```
 */
class SnapkitImageURL(private val organizationName: String) {

    /**
     * Generate Snapkit image proxy URL
     *
     * @param path S3 path (e.g., "project/path/to/image.jpg") - for direct S3 access
     * @param url Original image URL - for external CDN proxy
     * @param transform Image transformation options
     * @return Complete image proxy URL
     */
    fun build(path: String? = null, url: String? = null, transform: TransformOptions? = null): String {
        // Validate parameters
        require(path != null || url != null) {
            "Either 'path' or 'url' parameter must be provided"
        }

        require(!(path != null && url != null)) {
            "Cannot use both 'path' and 'url' parameters simultaneously"
        }

        val transformString = transform?.let { buildTransformString(it) } ?: ""

        // S3 direct access
        if (path != null) {
            val baseUrl = "https://$organizationName-cdn.snapkit.studio/$path"
            return if (transformString.isNotEmpty()) {
                "$baseUrl?transform=$transformString"
            } else {
                baseUrl
            }
        }

        // External CDN proxy
        val baseUrl = "https://$organizationName.snapkit.dev/image"
        val uriBuilder = Uri.parse(baseUrl).buildUpon()
        uriBuilder.appendQueryParameter("url", url)

        if (transformString.isNotEmpty()) {
            uriBuilder.appendQueryParameter("transform", transformString)
        }

        return uriBuilder.build().toString()
    }

    private fun buildTransformString(options: TransformOptions): String {
        val parts = mutableListOf<String>()

        // Numeric/string value parameters
        options.w?.let { parts.add("w:$it") }
        options.h?.let { parts.add("h:$it") }
        options.fit?.let { parts.add("fit:$it") }
        options.format?.let { parts.add("format:$it") }
        options.rotation?.let { parts.add("rotation:$it") }
        options.blur?.let { parts.add("blur:$it") }
        options.dpr?.let { parts.add("dpr:$it") }
        options.quality?.let { parts.add("quality:$it") }

        // Boolean parameters
        if (options.grayscale == true) parts.add("grayscale")
        if (options.flip == true) parts.add("flip")
        if (options.flop == true) parts.add("flop")

        // extract parameter
        options.extract?.let {
            parts.add("extract:${it.x}-${it.y}-${it.width}-${it.height}")
        }

        return parts.joinToString(",")
    }
}

File: swift/Sources/SnapkitImageURL/SnapkitImageURL.swift
import Foundation

/// Image transformation parameters
public struct TransformOptions {
    /// Image width (pixels)
    public var w: Int?
    /// Image height (pixels)
    public var h: Int?
    /// Resize method
    public var fit: Fit?
    /// Output format
    public var format: Format?
    /// Rotation angle (degrees)
    public var rotation: Int?
    /// Blur intensity (0.3-1000)
    public var blur: Int?
    /// Whether to convert to grayscale
    public var grayscale: Bool?
    /// Whether to flip vertically
    public var flip: Bool?
    /// Whether to flip horizontally
    public var flop: Bool?
    /// Region extraction
    public var extract: Extract?
    /// Device Pixel Ratio (1.0-4.0)
    public var dpr: Double?
    /// Image quality (1-100)
    public var quality: Int?

    public init(
        w: Int? = nil,
        h: Int? = nil,
        fit: Fit? = nil,
        format: Format? = nil,
        rotation: Int? = nil,
        blur: Int? = nil,
        grayscale: Bool? = nil,
        flip: Bool? = nil,
        flop: Bool? = nil,
        extract: Extract? = nil,
        dpr: Double? = nil,
        quality: Int? = nil
    ) {
        self.w = w
        self.h = h
        self.fit = fit
        self.format = format
        self.rotation = rotation
        self.blur = blur
        self.grayscale = grayscale
        self.flip = flip
        self.flop = flop
        self.extract = extract
        self.dpr = dpr
        self.quality = quality
    }

    /// Resize method
    public enum Fit: String {
        case contain
        case cover
        case fill
        case inside
        case outside
    }

    /// Output format
    public enum Format: String {
        case jpeg
        case png
        case webp
        case avif
    }

    /// Region extraction
    public struct Extract {
        public let x: Int
        public let y: Int
        public let width: Int
        public let height: Int

        public init(x: Int, y: Int, width: Int, height: Int) {
            self.x = x
            self.y = y
            self.width = width
            self.height = height
        }
    }
}

/// Snapkit image URL builder
public struct SnapkitImageURLBuilder {
    private let organizationName: String

    public init(organizationName: String) {
        self.organizationName = organizationName
    }

    /// Generate Snapkit image proxy URL
    ///
    /// - Parameters:
    ///   - path: S3 path (e.g., "project/path/to/image.jpg") - for direct S3 access
    ///   - url: Original image URL (CloudFront, etc.) - for external CDN proxy
    ///   - transform: Image transformation options
    /// - Returns: Complete image proxy URL
    ///
    /// # Example - S3 Direct Access (Recommended)
    /// ```swift
    /// let builder = SnapkitImageURLBuilder(organizationName: "my-org")
    /// let imageURL = builder.build(
    ///     path: "project/images/hero.jpg",
    ///     transform: TransformOptions(
    ///         w: 300,
    ///         h: 200,
    ///         fit: .cover,
    ///         format: .webp
    ///     )
    /// )
    /// // → "https://my-org-cdn.snapkit.studio/project/images/hero.jpg?transform=w:300,h:200,fit:cover,format:webp"
    /// ```
    ///
    /// # Example - External CDN Proxy (Optional)
    /// ```swift
    /// let imageURL = builder.build(
    ///     url: "https://cdn.cloudfront.net/image.jpg",
    ///     transform: TransformOptions(
    ///         w: 300,
    ///         h: 200,
    ///         fit: .cover,
    ///         format: .webp
    ///     )
    /// )
    /// // → "https://my-org.snapkit.dev/image?url=https%3A%2F%2F...&transform=w:300,h:200,fit:cover,format:webp"
    /// ```
    public func build(path: String? = nil, url: String? = nil, transform: TransformOptions? = nil) -> URL? {
        // Validate parameters
        guard path != nil || url != nil else {
            return nil
        }

        guard !(path != nil && url != nil) else {
            return nil
        }

        let transformString = transform.map { buildTransformString($0) } ?? ""

        // S3 direct access
        if let path = path {
            let baseUrl = "https://\(organizationName)-cdn.snapkit.studio/\(path)"
            if !transformString.isEmpty {
                return URL(string: "\(baseUrl)?transform=\(transformString)")
            }
            return URL(string: baseUrl)
        }

        // External CDN proxy
        var components = URLComponents(string: "https://\(organizationName).snapkit.dev/image")
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "url", value: url)
        ]

        if !transformString.isEmpty {
            queryItems.append(URLQueryItem(name: "transform", value: transformString))
        }

        components?.queryItems = queryItems
        return components?.url
    }

    private func buildTransformString(_ options: TransformOptions) -> String {
        var parts: [String] = []

        // Numeric/string value parameters
        if let w = options.w {
            parts.append("w:\(w)")
        }
        if let h = options.h {
            parts.append("h:\(h)")
        }
        if let fit = options.fit {
            parts.append("fit:\(fit.rawValue)")
        }
        if let format = options.format {
            parts.append("format:\(format.rawValue)")
        }
        if let rotation = options.rotation {
            parts.append("rotation:\(rotation)")
        }
        if let blur = options.blur {
            parts.append("blur:\(blur)")
        }
        if let dpr = options.dpr {
            parts.append("dpr:\(dpr)")
        }
        if let quality = options.quality {
            parts.append("quality:\(quality)")
        }

        // Boolean parameters
        if options.grayscale == true {
            parts.append("grayscale")
        }
        if options.flip == true {
            parts.append("flip")
        }
        if options.flop == true {
            parts.append("flop")
        }

        // extract parameter
        if let extract = options.extract {
            parts.append("extract:\(extract.x)-\(extract.y)-\(extract.width)-\(extract.height)")
        }

        return parts.joined(separator: ",")
    }
}
